### DISCUSSION TITLE
Fetch requests sometimes stuck in pending state using WSL2

### DISCUSSION THREAD

**Discussion by 39digits** [remix-run/remix#1642] [Q&A] [answered]
I wanted to avoid creating an issue until I knew more about what is happening in hopes someone else here has any ideas.

### Behaviour Description
Fetch requests get stuck in a pending state when running a Remix app under WSL2.

When the fetch request shows as pending the Remix app sometimes responds after a few clicks (and many cancelled fetch requests) while sometimes it requires a hard refresh.  When a request gets stuck in the pending state it never resolves by itself.

The pending state is easily seen in the Network tab of the browser dev tools.

### System Information

- Windows Version: Windows 11 Pro 21H2 (22000.438)
- WSL2 version: 0.51.2.0
- Kernel version: 5.10.81.1
- Linux Distribution:  Ubuntu 20.04
- Remix versions tested:  1.1.1 & 1.1.3
- Browsers used: Chrome, Edge

### Remix App Information
I've tried with a few different Remix apps and it happens on each.

The simplest setup is to replicate the Developer Blog tutorial example up to the Dynamic Route Params (so omitting the Admin parts) and creating a few extra markdown posts for some added variety.  The only change I made was to add a `<Link to="/posts">Posts</Link>` to the `$slug.tsx` file to easily and quickly navigate back to the Post listing.

I have tried the exact same app on macOS and never experience any issues.  I also don't have any similar issues with Next or Gatsby applications running in the same WSL2 distro so it feels specific to Remix on WSL2.

### How To Replicate
A good way to try replicate it is to click through links quite quickly and then leave the browser tab for a few seconds before clicking on a link.

In the Developer Blog I would click from the post listing to a blog back to `/posts` and back to another blog post a few times.  Leaving it for a few seconds (10+) and the next click will (possibly!) get stuck in a pending state.

Worth noting that this tactic doesn't always work but it's the best way I've found to replicate the issue locally.

---------
Edit: 25/01/2022 - I just realised that the information shown for the development URL changed from http://localhost:3000 (Remix 1.1.1) to showing the actual IP address on port 3000 (1.1.3).

Upon testing against both URLS, I see the issue above only occurs when browsing the development site using http://localhost:3000.  To be explicit a fetch request does not get stuck in a pending state when visiting the dev build using the IP address.

Is there something Remix does differently that causes issues using http://localhost:3000 compared to other frameworks? I've been so used to using localhost for years now.

**Comment by ccssmnn** [comment]
I don't have a Windows machine for testing, but can you tell if the server is being reached?

You can check both the logs of the remix dev server & the network tab of your browser. There you should find out where the request gets stuck.

**Reply by 39digits** [reply]
Good point.

I forgot to mention that the Remix logs do not show any hits for that endpoint when it shows a stuck pending state.

Meanwhile in the browser the Network Timing shows it stalls on the Connection Start -> Initial Connection.

**Reply by ccssmnn** [reply]
If the request URLs look good, but the connection cannot be established, this must be a problem with routing in WSL.

From what you say, remix is making requests in the frontend, but they don't make it to the server.

**Reply by 39digits** [reply]
I need to do some digging - I'm now curious why apps using Remix are the only ones affected by these intermittent issues using localhost while Next, Gatsby, etc are all 100% fine.

**Reply by 39digits** [reply]
Actually - I just created a quick Express "Hello World" app and it seems that's the culprit as some localhost requests get stuck in that Pending state.  I know remix-serve uses that under the hood so it's not a Remix-specific thing at least.

**Reply by ccssmnn** [reply]
Okay. Would be interesting, what the difference is between this and the requests from next/gatsby. Because it must be a problem with the network between Windows and WSL.

**Reply by 39digits** [reply]
One step closer - Next app started to do the same (Gatsby 100% unaffected) so it's absolutely a WSL2 issue.  And it looks to be specifically around the keep-alive timeout.

The default keep-alive timeout is 5seconds which is why everything works when I'm clicking around the Remix app without a delay but if I wait a short while it gets stuck.

In the test Express app I explicitly set the Node HTTP server keepAliveTimeout setting to 0 as a test and everything worked without issue.

**Reply by lucioladen** [reply]
@39digits  I'm struggling searching how can I setup keepAliveTimeout in Remix and I can't find it anywhere. Are you using built-in remix server or with an express adapter? (can't find any kind of server configuration in Remix for that matter..)

**Reply by 39digits** [reply]
It's been a while since I looked at this one but it's layered...like an onion...and it was down to a bug in WSL2 and the default Node setting.

I was using Remix with the Remix App Server which used Express which is running on Node.  When I was experiencing the issue it wasn't isolated to just Remix but any Express-based solution (including a very simple Express app).  Changing the [keepAliveTimeout setting](https://nodejs.org/api/http.html#http_server_keepalivetimeout) from the default made a difference during testing.

I raised a [bug with the WSL2 team](https://github.com/microsoft/WSL/issues/7982) at the time and they fixed it pretty quickly.  I haven't had an issue since then.

**Comment by 39digits** [comment, marked as answer]
I've experienced this under other Express-based web apps so it's not isolated to Remix.  This seems to be a keep alive issue with the latest update to WSL2 and not a Remix issue so I'll mark this discussion answered.  

I've opened an [issue under the WSL repo](https://github.com/microsoft/WSL/issues/7982) if anybody else has the same issue and wants to follow progress.
